diff --git a/thrift/lib/CMakeLists.txt b/thrift/lib/CMakeLists.txt
index 101f2d1ddc..e4c1e7ec22 100644
--- a/thrift/lib/CMakeLists.txt
+++ b/thrift/lib/CMakeLists.txt
@@ -28,7 +28,6 @@ set(LIB_DIRS
   thrift
   cpp
   cpp2
-  py3
 )
 foreach(dir ${LIB_DIRS})
   install(DIRECTORY ${dir} "${CMAKE_CURRENT_BINARY_DIR}/${dir}"
diff --git a/thrift/lib/py3/CMakeLists.txt b/thrift/lib/py3/CMakeLists.txt
index 9929cf0cc1..68791d48a6 100644
--- a/thrift/lib/py3/CMakeLists.txt
+++ b/thrift/lib/py3/CMakeLists.txt
@@ -114,7 +114,7 @@ foreach(_mod ${cpp_tgts})
   )
 
   if(${_f} STREQUAL "types")
-    python_add_module(${_module_name} "${_cxx}" "enums.cpp")
+    python_add_module(${_module_name} "${_cxx}")
   else()
     python_add_module(${_module_name} "${_cxx}")
   endif()
